### DISCUSSION TITLE
Trying to get Mac universal2 support from spec file

### DISCUSSION THREAD

**Discussion by Moosems** [pyinstaller/pyinstaller#8429] [Help] [answered]
I tried comparing `pyinstaller --target-arch universal2 main.py` and the spec file it created against my current spec file and couldn't find out what needed to be changed to get universal2 builds. Any thoughts?

Only thing I found different that I thought looked related was this:
```diff
exe = EXE(
    pyz,
    a.scripts,
-    [],
+    [('O', None, 'OPTION')],
```

**Comment by rokm** [comment, marked as answer]
It should set `target_arch='universal2'` argument to `EXE`.

**Reply by Moosems** [reply]
@rokm Now I'm getting the is not a fat binary from the `PIL` package, any thoughts on how to get around this?
```
PyInstaller.utils.osx.IncompatibleBinaryArchError: /Users/runner/work/DIP-dev-Tukaan/DIP-dev-Tukaan/venv/lib/python3.11/site-packages/PIL/_imagingtk.cpython-311-darwin.so is not a fat binary!
```

Thanks for all your help btw, it is very much appreciated!

**Reply by rokm** [reply]
You can only build fat binary if all python extensions are fat binaries.

See https://pyinstaller.org/en/stable/feature-notes.html#macos-multi-arch-support

**Reply by rokm** [reply]
If upstream provides only single-arch wheels, you can try using `delocate-fuse` from https://pypi.org/project/delocate/ to create universal2 wheel.

**Reply by Moosems** [reply]
Yes, I saw that page before but I would've thought that PIL would fit that description. PIL seems to only do single arch wheels and for now I'll just use an Intel Mac runner until I actually need to swap over and by then I'll probably use an m series chip.

Regardless, I think this is my last question for now: why can't PyInstaller to find the dylib here?
<img width="1020" alt="Screenshot 2024-04-26 at 1 13 46 PM" src="https://github.com/pyinstaller/pyinstaller/assets/95927277/8f3256e0-cc5b-4d94-9539-74a46f661d04">

```
Last login: Fri Apr 26 13:14:00 on ttys002
/Users/Moosems/Desktop/DIP-dev-Tukaan/dist/DIP/DIP ; exit;
Moosems-MacBook-Pro.local $ /Users/Moosems/Desktop/DIP-dev-Tukaan/dist/DIP/DIP ; exit;
Traceback (most recent call last):
  File "main.py", line 4, in <module>
  File "DIP/__main__.py", line 44, in main
  File "DIP/__main__.py", line 10, in __init__
  File "tukaan/app.py", line 38, in __init__
  File "libtukaan/serif.py", line 22, in init
  File "tukaan/_tcl.py", line 94, in load_dll
_tkinter.TclError: dlopen(/Users/Moosems/Desktop/DIP-dev-Tukaan/dist/DIP/_internal/libtukaan/mac/libserif_x64.dylib, 0x0006): tried: '/Users/Moosems/Desktop/DIP-dev-Tukaan/dist/DIP/_internal/libtukaan/mac/libserif_x64.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/Moosems/Desktop/DIP-dev-Tukaan/dist/DIP/_internal/libtukaan/mac/libserif_x64.dylib' (no such file), '/Users/Moosems/Desktop/DIP-dev-Tukaan/dist/DIP/_internal/libtukaan/mac/libserif_x64.dylib' (no such file)
[31290] Failed to execute script 'main' due to unhandled exception: dlopen(/Users/Moosems/Desktop/DIP-dev-Tukaan/dist/DIP/_internal/libtukaan/mac/libserif_x64.dylib, 0x0006): tried: '/Users/Moosems/Desktop/DIP-dev-Tukaan/dist/DIP/_internal/libtukaan/mac/libserif_x64.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/Moosems/Desktop/DIP-dev-Tukaan/dist/DIP/_internal/libtukaan/mac/libserif_x64.dylib' (no such file), '/Users/Moosems/Desktop/DIP-dev-Tukaan/dist/DIP/_internal/libtukaan/mac/libserif_x64.dylib' (no such file)
[31290] Traceback:
Traceback (most recent call last):
  File "main.py", line 4, in <module>
  File "DIP/__main__.py", line 44, in main
  File "DIP/__main__.py", line 10, in __init__
  File "tukaan/app.py", line 38, in __init__
  File "libtukaan/serif.py", line 22, in init
  File "tukaan/_tcl.py", line 94, in load_dll
_tkinter.TclError: dlopen(/Users/Moosems/Desktop/DIP-dev-Tukaan/dist/DIP/_internal/libtukaan/mac/libserif_x64.dylib, 0x0006): tried: '/Users/Moosems/Desktop/DIP-dev-Tukaan/dist/DIP/_internal/libtukaan/mac/libserif_x64.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/Moosems/Desktop/DIP-dev-Tukaan/dist/DIP/_internal/libtukaan/mac/libserif_x64.dylib' (no such file), '/Users/Moosems/Desktop/DIP-dev-Tukaan/dist/DIP/_internal/libtukaan/mac/libserif_x64.dylib' (no such file)


Saving session...
...copying shared history...
...saving history...truncating history files...
...completed.

[Process completed]
```

**Reply by rokm** [reply]
> Regardless, I think this is my last question for now: why can't PyInstaller to find the dylib here?

Presumably because line 5 in your hook overwrites the results of line 4?

**Reply by Moosems** [reply]
Thank you :)

**Reply by Moosems** [reply]
Can I compile for x86_64 from M1?

**Reply by rokm** [reply]
> Can I compile for x86_64 from M1?

If you python environment is fully `universal2`, you can create any kind of build (`x86_64`, `arm64`, `universal2`) on either `x86_64` or `arm64` machine. 

If not, you can create `x86_64` build on `arm64` machine by running PyInstaller in `x86_64` python environment through Rosetta.
